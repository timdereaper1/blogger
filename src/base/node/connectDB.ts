import { MongoClient } from 'mongodb';

let cached: { conn: MongoClient | null; promise: Promise<MongoClient> | null };

if (!cached) {
	cached = global.mongo = { conn: null, promise: null };
}

const environment = {
	development: {
		database: 'blogger',
		url: process.env.MONGO_URL,
	},
	test: {
		database: 'test_blogger',
		url: process.env.MONGO_URL,
	},
	production: {
		database: 'blogger',
		url: '',
	},
};

export function getDatabaseConfig() {
	const config = environment[process.env.NODE_ENV];
	return config;
}

export async function createDatabaseConnection() {
	if (cached.conn) return cached.conn;

	if (!cached.promise) {
		const config = getDatabaseConfig();
		cached.promise = MongoClient.connect(config.url);
		cached.conn = await cached.promise;
		return cached.conn;
	}
}
