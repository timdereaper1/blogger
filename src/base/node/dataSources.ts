import { Db } from 'mongodb';
import { UsersRepository, UsersRepositoryInterface } from './repositories/usersRepository';

export interface DataSourcesInterface {
	users: UsersRepositoryInterface;
}

export function DataSources(db: Db): DataSourcesInterface {
	const users = UsersRepository(db);

	return Object.freeze<DataSourcesInterface>({
		users,
	});
}
