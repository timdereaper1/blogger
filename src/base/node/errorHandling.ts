import { ApolloError } from 'apollo-server-micro';
import { processErrorToErrorLogs } from './logging';

export function processRequestError(error: any, defaultMessage?: string | ApolloError) {
	if (error.isCustom && defaultMessage) throwApolloErrorOrErrorMessage(defaultMessage);
	if (error.isCustom) throw error;
	processErrorToErrorLogs(error);
	throw new Error('Internal server error');
}

export function throwApolloErrorOrErrorMessage(error: string | ApolloError) {
	if (typeof error === 'string') throw new Error(error);
	throw error;
}
