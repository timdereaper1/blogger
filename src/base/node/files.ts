import fs from 'fs';
import path from 'path';

export function createDirectory(directoryPath: string) {
	const fullDirectoryPath = path.join(process.cwd(), directoryPath);
	const stats = fs.statSync(fullDirectoryPath);
	if (!stats.isDirectory()) fs.mkdirSync(fullDirectoryPath);
	return fullDirectoryPath;
}

export function createFile(filePath: string) {
	const stats = fs.statSync(filePath);
	if (!stats.isFile()) fs.writeFileSync(filePath, '');
	return filePath;
}

export function appendToFile(filePath: string, data: string) {
	const file = createFile(filePath);
	fs.appendFileSync(file, data);
	return true;
}
