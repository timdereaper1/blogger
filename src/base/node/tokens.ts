import { decode, sign, verify } from 'jsonwebtoken';

export function createAuthenticationToken(args: string) {
	return sign(args, 'secret-key-used-for-tokens');
}

export function getAuthenticatedUserIdInToken(token: string): string | null {
	const isTokenValid = verifyAuthenticationToken(token);
	if (!isTokenValid) return null;
	const decoded = decode(token);
	if (!decoded) return null;
	if (typeof decoded === 'object' && 'id' in decoded) return decoded.id;
	return decoded as any;
}

export function createExpiryAuthenticationToken(args: string, expiresIn: string | number = '1h') {
	return sign({ id: args }, 'secret-key-used-for-tokens', {
		expiresIn,
	});
}

export function verifyAuthenticationToken(token: string) {
	try {
		verify(token, 'secret-key-used-for-tokens');
		return true;
	} catch (error) {
		return false;
	}
}
