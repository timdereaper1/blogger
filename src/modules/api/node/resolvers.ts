import {
	signUpAccountMutation,
	verifyAndSendPasswordResetEmailMutation,
	verifyLoginCredentialsMutation,
} from 'src/modules/authentication/node/authenticationResolvers';

export const resolvers = {
	Query: {
		hello() {
			return 'Hello world';
		},
	},
	Mutation: {
		verifyCredentials: verifyLoginCredentialsMutation,
		signUpAccount: signUpAccountMutation,
		verifyAndSendPasswordResetEmail: verifyAndSendPasswordResetEmailMutation,
	},
};
