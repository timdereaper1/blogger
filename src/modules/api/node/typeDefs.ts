import { gql } from 'apollo-server-micro';

export const typeDefs = gql`
	type User {
		id: String!
		name: String!
		email: String!
		token: String
		privileges: [String!]!
	}

	type Success {
		message: String!
		success: Boolean!
	}

	input UserLoginCredentials {
		email: String!
		password: String!
	}

	input UserSignUpCredentials {
		password: String!
		email: String!
		name: String!
		privileges: [String!]!
	}

	input UserPasswordResetCredentials {
		email: String!
	}

	type Query {
		hello: String
	}

	type Mutation {
		verifyCredentials(data: UserLoginCredentials!): User
		signUpAccount(data: UserSignUpCredentials!): User
		verifyAndSendPasswordResetEmail(data: UserPasswordResetCredentials!): Success
	}
`;
