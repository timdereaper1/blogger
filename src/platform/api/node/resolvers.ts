import { loginGraphQLMutationResolvers } from 'src/platform/login/node/loginGraphQLResolver';

export const resolvers = {
	Query: {
		hello() {
			return 'Hello world';
		},
	},
	Mutation: Object.assign({}, loginGraphQLMutationResolvers),
};
