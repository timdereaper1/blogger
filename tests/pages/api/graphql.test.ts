import { ApolloServer } from 'apollo-server-micro';
import * as argon from 'argon2';
import type { DBUser } from 'src/base/node/repositories/types';
import { createAuthenticationToken } from 'src/base/node/tokens';
import { resolvers } from 'src/modules/api/node/resolvers';
import { typeDefs } from 'src/modules/api/node/typeDefs';
import {
	SIGN_IN_ACCOUNT_MUTATION,
	SIGN_UP_ACCOUNT_MUTATION,
	VERIFY_AND_SEND_PASSWORD_RESET_EMAIL_MUTATION,
} from 'src/modules/authentication/web/schemas';

jest.mock('src/base/node/services/mailing');
jest.mock('src/base/node/tokens');
const mockedToken = createAuthenticationToken as jest.MockedFunction<
	typeof createAuthenticationToken
>;

describe('graphql', () => {
	const sources = {
		users: {
			findByEmail: jest.fn(),
			insert: jest.fn(),
		},
	};

	const server = new ApolloServer({
		typeDefs,
		resolvers,
		dataSources: (): any => sources,
	});

	const password = '1234567';
	const email = 'john.doe@gmail.com';
	const name = 'John Doe';
	const token =
		'eyJhbGciOiJIUzI1NiJ9.OTM4NGRFZTM4ckVnZjkzZjAybkdmOTM.P2UQL5NOeeRzWqFUk8SggYtAvMi73wmsh3DGUzJuAdg';

	async function getDBUser(): Promise<DBUser> {
		return {
			email,
			_id: '9384dEe38rEgf93f02nGf93',
			name,
			password: await argon.hash(password),
			privileges: ['user'],
		};
	}

	describe('sign in', () => {
		it('should sign in user with correct credentials', async () => {
			const user = await getDBUser();
			sources.users.findByEmail.mockResolvedValueOnce(user);
			mockedToken.mockReturnValueOnce(token);
			const response = await server.executeOperation({
				query: SIGN_IN_ACCOUNT_MUTATION,
				variables: { data: { email, password } },
			});
			expect(response).toMatchSnapshot();
		});
	});

	describe('sign up', () => {
		it('should create a new account when user is signing up', async () => {
			const user = await getDBUser();
			sources.users.insert.mockResolvedValueOnce(user);
			mockedToken.mockReturnValueOnce(token);
			const response = await server.executeOperation({
				query: SIGN_UP_ACCOUNT_MUTATION,
				variables: { data: { email, name, password, privileges: ['user'] } },
			});
			expect(response).toMatchSnapshot();
		});
	});

	describe('send password reset email', () => {
		it('should send an email to user to reset their password', async () => {
			const user = await getDBUser();
			sources.users.findByEmail.mockResolvedValueOnce(user);
			const response = await server.executeOperation({
				query: VERIFY_AND_SEND_PASSWORD_RESET_EMAIL_MUTATION,
				variables: { data: { email } },
			});
			expect(response).toMatchSnapshot();
		});
	});
});
