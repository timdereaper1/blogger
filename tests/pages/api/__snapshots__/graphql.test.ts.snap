// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql send password reset email should send an email to user to reset their password 1`] = `
Object {
  "data": Object {
    "verifyAndSendPasswordResetEmail": Object {
      "message": "Reset password email has been sent to John Doe",
      "success": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`graphql sign in should sign in user with correct credentials 1`] = `
Object {
  "data": Object {
    "verifyCredentials": Object {
      "email": "john.doe@gmail.com",
      "id": "9384dEe38rEgf93f02nGf93",
      "name": "John Doe",
      "privileges": Array [
        "user",
      ],
      "token": "eyJhbGciOiJIUzI1NiJ9.OTM4NGRFZTM4ckVnZjkzZjAybkdmOTM.P2UQL5NOeeRzWqFUk8SggYtAvMi73wmsh3DGUzJuAdg",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`graphql sign up should create a new account when user is signing up 1`] = `
Object {
  "data": Object {
    "signUpAccount": Object {
      "email": "john.doe@gmail.com",
      "id": "9384dEe38rEgf93f02nGf93",
      "name": "John Doe",
      "privileges": Array [
        "user",
      ],
      "token": "eyJhbGciOiJIUzI1NiJ9.OTM4NGRFZTM4ckVnZjkzZjAybkdmOTM.P2UQL5NOeeRzWqFUk8SggYtAvMi73wmsh3DGUzJuAdg",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
